Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/Internal/CustomerCancellationDues.hs

```diff
@@ -0,0 +1,151 @@
+      disputeAmount = min (transporterConfig.cancellationFee * fromIntegral disputeChancesLeft) riderDetails.cancellationDues
+      disputeChances = disputeAmount `div` transporterConfig.cancellationFee
+  QRD.updateDisputeChancesUsed riderDetails.id (riderDetails.disputeChancesUsed + fromIntegral disputeChances) >> QRD.updateCancellationDues riderDetails.id (riderDetails.cancellationDues - disputeAmount)
+  pure Success
+
+getCancellationDuesDetails ::
+  ( MonadFlow m,
+    EsqDBFlow m r,
+    CacheFlow m r,
+    EncFlow m r
+  ) =>
+  Id Merchant ->
+  Maybe Text ->
+  CancellationDuesReq ->
+  m CancellationDuesDetailsRes
+getCancellationDuesDetails merchantId apiKey CancellationDuesReq {..} = do
+  merchant <- QM.findById merchantId >>= fromMaybeM (MerchantNotFound merchantId.getId)
+  unless (Just merchant.internalApiKey == apiKey) $
+    throwError $ AuthBlocked "Invalid BPP internal api key"
+  numberHash <- getDbHash customerMobileNumber
+  riderDetails <- QRD.findByMobileNumberHashAndMerchant numberHash merchant.id >>= fromMaybeM (InternalError "Rider Details Not Found")
+  return $ CancellationDuesDetailsRes {customerCancellationDues = riderDetails.cancellationDues, disputeChancesUsed = riderDetails.disputeChancesUsed}
+
+customerCancellationDuesSync ::
+  ( MonadFlow m,
+    EsqDBFlow m r,
+    CacheFlow m r,
+    EncFlow m r
+  ) =>
+  Id Merchant ->
+  Maybe Text ->
+  CustomerCancellationDuesSyncReq ->
+  m APISuccess
+customerCancellationDuesSync merchantId apiKey req = do
+  merchant <- QM.findById merchantId >>= fromMaybeM (MerchantNotFound merchantId.getId)
+  unless (Just merchant.internalApiKey == apiKey) $
+    throwError $ AuthBlocked "Invalid BPP internal api key"
+  numberHash <- getDbHash req.customerMobileNumber
+
+  when (isJust req.cancellationCharges && isJust req.disputeChancesUsed) $ do
+    throwError $ InvalidRequest "Either of the two , Due Amount or Dispute Chances has to be Null"
+  merchantOperatingCity <- CQMM.findByMerchantIdAndCity merchantId merchant.city >>= fromMaybeM (MerchantOperatingCityNotFound $ "merchant-Id-" <> merchant.id.getId <> "-city-" <> show merchant.city)
+  transporterConfig <- SCT.findByMerchantOpCityId merchantOperatingCity.id >>= fromMaybeM (TransporterConfigNotFound merchantOperatingCity.id.getId)
+  riderDetails <- QRD.findByMobileNumberHashAndMerchant numberHash merchant.id >>= fromMaybeM (InternalError "Rider Details Not Found")
+  case (req.cancellationCharges, req.disputeChancesUsed) of
+    (Just amountPaid, Nothing) -> do
+      when (amountPaid > riderDetails.cancellationDues || amountPaid < 0) $ do
+        throwError $ InvalidRequest "Limits not met for the input request"
+      QRD.updateCancellationDues riderDetails.id (riderDetails.cancellationDues - amountPaid)
+      when (req.paymentMadeToDriver) $ do
+        booking <- (QBooking.findLastCancelledByRiderId riderDetails.id) >>= fromMaybeM (BookingDoesNotExist riderDetails.id.getId) . listToMaybe
```

Comment:
can we fetch one row from DB instead of fetching multiple rows and then use `listToMaybe`?

Reply:
Limit is 1 so it will give list of one booking only

Reply:
Then `findLastCancelledByRiderId` should return `Maybe`, because we know that there is only one Booking, not list.
Just move `listToMaybe` into query
`findLastCancelledByRiderId :: (MonadFlow m, EsqDBFlow m r, CacheFlow m r) => Id RiderDetails -> m (Maybe Booking)`